ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -16,27 +22,111 @@ import           Data.Maybe (isJust)
 
+-- | Different queries supported by the ledger for all block types, indexed
+-- by the result type.
+data Query blk result = BlockQuery (BlockQuery blk result)
+
+instance (ShowProxy (BlockQuery blk)) => ShowProxy (Query blk) where
+  showProxy (Proxy :: Proxy (Query blk)) = "Query (" ++ showProxy (Proxy @(BlockQuery blk)) ++ ")"
+
+instance (ShowQuery (BlockQuery blk)) => ShowQuery (Query blk) where
+  showResult (BlockQuery blockQuery) = showResult blockQuery
+
+instance Eq (SomeSecond BlockQuery blk) => Eq (SomeSecond Query blk) where
+  SomeSecond (BlockQuery blockQueryA) == SomeSecond (BlockQuery blockQueryB)
+    = SomeSecond blockQueryA == SomeSecond blockQueryB
+
+instance Show (SomeSecond BlockQuery blk) => Show (SomeSecond Query blk) where
+  show (SomeSecond (BlockQuery blockQueryA)) = "BlockQuery " ++ show (SomeSecond blockQueryA)
+
+instance ( SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+         , SupportedNetworkProtocolVersion blk
+         ) => SerialiseNodeToClient blk (SomeSecond Query blk) where
+  encodeNodeToClient  codecConfig version (SomeSecond query)
+    | not (isQuerySerialisationSupported (Proxy @blk) version)
+    = error "encode failure: Serializing Query is not supported"
+    | otherwise
+    = case query of
+      BlockQuery blockQuery -> encodeTag 0 <> encodeNodeToClient codecConfig version (SomeSecond blockQuery)
+
+  decodeNodeToClient codecConfig version
+    | not (isQuerySerialisationSupported (Proxy @blk) version)
+    = fail "decode failure: Deserializing Query is not supported"
+    | otherwise
+    = do
+      tag <- decodeTag
+      case tag of
+        0 -> decodeNodeToClient codecConfig version
+        _ -> fail $ "SomeSecond Query blk: unknown tag " ++ show tag
+
+instance ( SerialiseResult blk (BlockQuery blk)
+         , SupportedNetworkProtocolVersion blk
+         ) => SerialiseResult blk (Query blk) where
+  encodeResult codecConfig blockNodeToClientVersion query result
+    | not (isQuerySerialisationSupported (Proxy @blk) blockNodeToClientVersion)
+    = error "encode failure: Serializing Query is not supported"
+    | otherwise
+    = case query of
+        BlockQuery blockQuery ->
+          encodeTag 0
+          <> encodeResult codecConfig blockNodeToClientVersion blockQuery result
+
+  decodeResult codecConfig blockNodeToClientVersion query
+    = do
+      tag <- decodeTag
+      case query of
+        BlockQuery blockQuery
+          | tag /= 0 -> fail $ "Query blk: Expected tag 0 but got " ++ show tag
+          | otherwise -> decodeResult codecConfig blockNodeToClientVersion blockQuery
+
+isQuerySerialisationSupported :: (SupportedNetworkProtocolVersion blk) => Proxy blk -> BlockNodeToClientVersion blk -> Bool
+isQuerySerialisationSupported blkP blockN2CV = any (>= NodeToClientV_9) supportedN2CVs
+  where
+    supportedN2CVs =
+      [ node2CV
+      | (node2CV, blockN2CV') <- Map.toList (supportedNodeToClientVersions blkP)
+      , blockN2CV == blockN2CV'
+      ]
```

Comment:
@nfrisby I've added this version check, but things are a bit awkeward as the check depends on the `NodeToClientVersion` rather than the `BlockNodeToClientVersion blk`, so I've ended up having to add a `SupportedNetworkProtocolVersion blk` constraint in various places (this is all done in the latest commit). Any thoughts on this or if there is a better way of doing this?

Reply:
This is resolved in the latest version.